### PR TITLE
Feature/add x content security policy

### LINKED ISSUE #6: Allow configuration for X-Content-Security-Policy
**This is a duplicate of https://github.com/GaProgMan/OwaspHeaders.Core/issues/47 to track the progress here as well.**

This is an issue to discuss the Pull Request I am to file soon.

There are different sources about if it is a good idea to add the X-Content-Security-Policy-header. 

https://content-security-policy.com/ says ` Note: It is known that having both Content-Security-Policy and X-Content-Security-Policy or X-Webkit-CSP causes unexpected behaviours on certain versions of browsers. Please avoid using deprecated X-* headers.`

I did not find out what this "unexpected behaviour" is and sources like [stackexchange](https://security.stackexchange.com/questions/191455/whats-the-alternative-of-content-security-policy-csp-header-in-internet-explo) or [stackoverflow](https://stackoverflow.com/questions/42937146/content-security-policy-does-not-work-in-internet-explorer-11) tell that this header is the only way to make Internet Explorer a bit more secure, although not all properties work.

Well-respected products like [IdentityServer](https://github.com/IdentityServer/IdentityServer4.Demo/blob/master/src/IdentityServer4Demo/Quickstart/SecurityHeadersAttribute.cs#L42) also set this header in their Quickstart-Guide, to the same value as the normal Content-Security-Policy header.

In my understanding it does not hurt to set it to the same value and by adding "sandbox" and optionally ["allow-scripts" "allow-forms"](https://blogs.msdn.microsoft.com/ie/2011/07/14/defense-in-depth-locking-down-mash-ups-with-html5-sandbox/) it should help with IE. I did not find a good source about "allow-same-origin". 

I am currently working on adding an optional config.UseXContentSecurityPolicy - property (set to false by default) that will just put the content of the Content-Security-Policy-Header to the X-Content-Security-Policy-Header as well. Also on the TODO is to test that I can add "sandbox allow-scripts allow-same-origin allow-forms" to this header.

I am not quite sure if this is the right route to go, so if anyone has any better ideas feel free to add those here. 

Pull request will come soon.

### PR DESCRIPTION
**Pull request by Compufreak345**
fix #6

### CODE REVIEW

File: src/SecureHeadersMiddleware.cs [hunk 1]
```diff
@@ -64,6 +64,12 @@ public async Task Invoke(HttpContext httpContext)
                     _config.ContentSecurityPolicyConfiguration.BuildHeaderValue());
             }
 
+            if (_config.UseXContentSecurityPolicy)
+            {
+                httpContext.TryAddHeader(Constants.XContentSecurityPolicyHeaderName,
```
Comment (paule96) [inline src/SecureHeadersMiddleware.cs]: if this is write if the tryaddHeader fails we get nothing? (maybe a log warning or error would be nice)

File: src/Constants.cs [hunk 2]
```diff
@@ -12,6 +12,8 @@ public static class Constants
 
         public static readonly string ContentSecurityPolicyHeaderName = "Content-Security-Policy";
 
+        public static readonly string XContentSecurityPolicyHeaderName = "X-Content-Security-Policy";
```
[the remark this answers is not in the document]
Comment (Compufreak345) [reply src/Constants.cs]: It's probably there for a reason, I just followed the initial implementation. But yeah, we can check in another branch.

File: src/Models/ContentSecurityPolicySandBox.cs [hunk 3]
```diff
@@ -1,46 +1,63 @@
...
 {
     public class ContentSecurityPolicySandBox : IConfigurationBase
     {
-        protected ContentSecurityPolicySandBox() {}
+        protected ContentSecurityPolicySandBox() { }
 
-        public CspSandboxType SandboxType { get; set; }
-        
-        public ContentSecurityPolicySandBox(CspSandboxType sandboxType)
+        public IList<CspSandboxType> SandboxTypes { get; set; }
```
[the remark this answers is not in the document]
Comment (Compufreak345) [reply src/Models/ContentSecurityPolicySandBox.cs]: After our last talk about params, do you still think this is needed?

File: src/Models/ContentSecurityPolicySandBox.cs [hunk 4]
```diff
@@ -1,46 +1,63 @@
...
         {
-            SandboxType = sandboxType;
+            SandboxTypes = sandboxType;
         }
 
         public string BuildHeaderValue()
         {
-            var returnStr = "sandbox";
-            switch (SandboxType)
+            var returnStr = new StringBuilder("sandbox");
```
[the remark this answers is not in the document]
Comment (Compufreak345) [reply src/Models/ContentSecurityPolicySandBox.cs]: Regarding to this benchmark from 2019:
https://michaelscodingspot.com/challenging-the-c-stringbuilder/

The Stringbuilder begins to be worth it at 4 concatenations, at least regarding performance what's probably the most important regarding web requests. As with the recent code changes for empty arrays we return an empty string we are around 3-6 concatenations, so I left it in.

File: src/SecureHeadersMiddleware.cs (hunk 1 above)
Comment (Compufreak345) [reply src/SecureHeadersMiddleware.cs]: I don't think we should change the default behavior of the package in this PR. If we want to fix this we should do this in another PR. But be aware that this middleware does not use any logger and I think we should keep it as tight as possible as it is used in every request.

File: src/Models/ContentSecurityPolicySandBox.cs [hunk 5]
```diff
@@ -1,46 +1,63 @@
...
+            foreach (var sandboxType in SandboxTypes)
```
[the remark this answers is not in the document]
Comment (Compufreak345) [reply src/Models/ContentSecurityPolicySandBox.cs]: Done.

File: src/Models/ContentSecurityPolicySandBox.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (paule96) [reply src/Models/ContentSecurityPolicySandBox.cs]: the change to IEnumerable yes. but not the null thing.

File: src/SecureHeadersMiddleware.cs (hunk 1 above)
Comment (paule96) [reply src/SecureHeadersMiddleware.cs]: okay

File: src/Models/ContentSecurityPolicySandBox.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (paule96) [reply src/Models/ContentSecurityPolicySandBox.cs]: Okay so we hope we get most of the time more then 4 concats.

**Comment by paule96** [comment]
go!

**Pull request merged**